### PR TITLE
Add new filter for empty fields

### PR DESCRIPTION
**Pull request by core23**
## Subject

This filter allows filtering for (non-)nullable fields.

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
-  Add new filter for empty fields
```

### CODE REVIEW

File: src/Filter/EmptyFieldFilter.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use Sonata\AdminBundle\Datagrid\ProxyQueryInterface;
+use Sonata\AdminBundle\Form\Type\Filter\DefaultType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+
+final class EmptyFieldFilter extends Filter
+{
+    /**
+     * Field is set.
+     */
+    public const TYPE_YES = 1;
```
Comment (OskarStark) [inline src/Filter/EmptyFieldFilter.php]: Couldn’t we use our own constants instead of yes and now with a separate meaning (phpdoc)?

File: docs/reference/filter_field_definition.rst [hunk 2]
```diff
@@ -44,6 +44,7 @@ For now, only `Doctrine ORM` filters are available:
 * ``Sonata\DoctrineORMAdminBundle\Filter\DateTimeFilter``: depends on the ``sonata_type_filter_datetime`` Form Type, renders a datetime field,
 * ``Sonata\DoctrineORMAdminBundle\Filter\DateTimeRangeFilter``: depends on the ``sonata_type_filter_datetime_range`` Form Type, renders a 2 datetime fields,
 * ``Sonata\DoctrineORMAdminBundle\Filter\ClassFilter``: depends on the ``sonata_type_filter_default`` Form type, renders a choice list field.
+* ``Sonata\DoctrineORMAdminBundle\Filter\EmptyFieldFilter``: depends on the ``sonata_type_filter_default`` Form type, renders a choice list field.
```
Comment (franmomu) [inline docs/reference/filter_field_definition.rst]:
```suggestion
* ``Sonata\DoctrineORMAdminBundle\Filter\EmptyFieldFilter``: depends on the ``Sonata\AdminBundle\Form\Type\Filter\DefaultType`` Form type, renders a choice list field.
```

File: src/Filter/EmptyFieldFilter.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            'choices' => [
+                'label_type_empty' => static::TYPE_EMPTY,
+                'label_type_not_empty' => static::TYPE_NOT_EMPTY,
+            ],
+        ]);
+    }
+
+    public function getRenderSettings()
+    {
+        return [DefaultType::class, [
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php]: I think you should create a SonataAdminBundle/Form/Type.
For example the `NumberFilter` use a `NumberType`.

Since there is multiple ORM Repository, this is a way to use same labels, constants, configuration, ...

File: src/Filter/EmptyFieldFilter.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+     * @param array|string|null $data
+     */
+    public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data): void
+    {
+        if (null === $data || !\is_array($data) || !\array_key_exists('value', $data)) {
+            return;
+        }
+
+        if ((int) $data['value'] === static::TYPE_NOT_EMPTY) {
+            $queryBuilder->andWhere(sprintf('%s.%s IS null', $alias, $field));
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php]: If I filter by TYPE_NOT_EMPTY, I expect the not null

Comment (VincentLanglet) [reply src/Filter/EmptyFieldFilter.php]: And would it possible to handle OTM, OTO, MTO, ... relations ?

If I have A human entity OTM with dog entity.
If I look for every human without dog (i.e. human.dogs.id null), or every dog with human without a name (i.e dog.human.name), could this filter be used ?

File: src/Filter/EmptyFieldFilter.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Filter/EmptyFieldFilter.php:69]: An "EmptyOperatorType" like this https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Form/Type/Operator/ContainsOperatorType.php in the SonataAdminBundle would be better.

This allow to be used by others PersistenceBundle.

File: tests/Filter/EmptyFieldFilterTest.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+
+final class EmptyFieldFilterTest extends TestCase
+{
+    public function testRenderSettings(): void
+    {
+        $filter = new EmptyFieldFilter();
+        $filter->initialize('field_name', [
+            'field_options' => ['class' => 'FooBar'],
+            ]);
```
Comment (phansys) [inline tests/Filter/EmptyFieldFilterTest.php:27]:
```suggestion
        ]);
```

File: src/Filter/EmptyFieldFilter.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        if ((int) $data['value'] === static::TYPE_EMPTY) {
+            $queryBuilder->andWhere(sprintf('%s.%s IS NULL', $alias, $field));
+        } elseif ((int) $data['value'] === static::TYPE_NOT_EMPTY) {
+            $queryBuilder->andWhere(sprintf('%s.%s IS NOT NULL', $alias, $field));
+        } else {
+            throw new \UnexpectedValueException('Invalid filter value given');
+        }
+
+        $this->active = null !== $data['value'];
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php:46]: Others filter of this bundle doesn't touch the `$this->active` property.

It's handle https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Filter/Filter.php#L62 for what I understand.

File: src/Filter/EmptyFieldFilter.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    public const TYPE_NOT_EMPTY = 2;
+
+    public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data): void
+    {
+        if (null === $data || !\is_array($data) || !\array_key_exists('value', $data)) {
+            return;
+        }
+
+        if ((int) $data['value'] === static::TYPE_EMPTY) {
+            $queryBuilder->andWhere(sprintf('%s.%s IS NULL', $alias, $field));
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php:39]: You have to use `applyWhere` to handle both OR filter and AND filter.

File: src/Filter/EmptyFieldFilter.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        } elseif ((int) $data['value'] === static::TYPE_NOT_EMPTY) {
+            $queryBuilder->andWhere(sprintf('%s.%s IS NOT NULL', $alias, $field));
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php:41]: You have to use `applyWhere` to handle both OR filter and AND filter.

File: src/Filter/EmptyFieldFilter.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            'choices' => [
+                'label_type_empty' => static::TYPE_EMPTY,
+                'label_type_not_empty' => static::TYPE_NOT_EMPTY,
+            ],
+        ]);
+    }
+
+    public function getRenderSettings()
+    {
+        return [DefaultType::class, [
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php:71]: WDYT about using the `yes/no` BooleanType ?

`Firstname: Yes` returns every not null values.
`Firstname: No` returns every null values.

**Comment by core23** [comment]
Thanks for the Feedback @VincentLanglet . This was just some very old code from my private project I want to share.

File: src/Filter/EmptyFieldFilter.php [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public function getParentAssociationMappings()
+    {
+        $mappings = $this->getOption('parent_association_mappings', []);
+
+        $fields = explode('.', $this->getFieldName());
+        for ($i = 0; $i < \count($fields) - 1; ++$i) {
+            $mappings[] = ['fieldName' => $fields[$i]];
+        }
```
Comment (phansys) [inline src/Filter/EmptyFieldFilter.php:78]:
```suggestion
        $fields = explode('.', $this->getFieldName(), -1);
        foreach ($fields as $field) {
            $mappings[] = ['fieldName' => $field];
        }
```

File: src/Filter/EmptyFieldFilter.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+     * @return string[]
+     */
+    protected function association(ProxyQueryInterface $queryBuilder, $data): array
+    {
+        $alias = $queryBuilder->entityJoin($this->getParentAssociationMappings());
+        $part = strrchr('.'.$this->getFieldName(), '.');
+        $fieldName = substr(false === $part ? $this->getFieldType() : $part, 1);
+
+        return [$alias, $fieldName];
+    }
```
Comment (VincentLanglet) [inline src/Filter/EmptyFieldFilter.php:96]: I'm curious. Why do you need this ?

There is nothing similar in `StringFilter`, `NumberFilter`, `ChoiceFilter` or `BooleanFilter`. 
What is different here ?

File: tests/Filter/EmptyFieldFilterTest.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Sonata\DoctrineORMAdminBundle\Tests\Filter;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\DoctrineORMAdminBundle\Filter\EmptyFieldFilter;
+use Sonata\Form\Type\BooleanType;
+
+final class EmptyFieldFilterTest extends TestCase
+{
+    public function testRenderSettings(): void
```
Comment (VincentLanglet) [inline tests/Filter/EmptyFieldFilterTest.php:22]: I think more tests could be great, at least two more tests, testing the query for the null case and the not null case.

Comment (core23) [reply tests/Filter/EmptyFieldFilterTest.php]: Class is now 100% covered

**Comment by phansys** [comment]
Thank you @core23!
